src/app/codegen-data-model/CodegenDataModel_Read.cpp

```diff
@@ -0,0 +1,370 @@
+                  request.path.mExpanded);
+
+    // ACL check for non-internal requests
+    if (!request.operationFlags.Has(InteractionModel::OperationFlags::kInternal))
+    {
+        ReturnErrorCodeIf(!request.subjectDescriptor.has_value(), CHIP_ERROR_INVALID_ARGUMENT);
+
+        Access::RequestPath requestPath{ .cluster = request.path.mClusterId, .endpoint = request.path.mEndpointId };
+        CHIP_ERROR err = Access::GetAccessControl().Check(*request.subjectDescriptor, requestPath,
+                                                          RequiredPrivilege::ForReadAttribute(request.path));
+        if (err != CHIP_NO_ERROR)
+        {
+            // Implementation of 8.4.3.2 of the spec for path expansion
+            if (request.path.mExpanded && (err == CHIP_ERROR_ACCESS_DENIED))
+            {
+                return CHIP_NO_ERROR;
+            }
+            return err;
+        }
+    }
+
+    auto metadata = FindAttributeMetadata(request.path);
+
+    // Explicit failure in finding a suitable metadata
+    if (const CHIP_ERROR * err = std::get_if<CHIP_ERROR>(&metadata))
+    {
+        VerifyOrDie(*err != CHIP_NO_ERROR);
+        return *err;
+    }
+
+    // Read via AAI
+    std::optional<CHIP_ERROR> aai_result;
+    if (const EmberAfCluster ** cluster = std::get_if<const EmberAfCluster *>(&metadata))
+    {
+        Compatibility::GlobalAttributeReader aai(*cluster);
+        aai_result = TryReadViaAccessInterface(request.path, &aai, encoder);
+    }
+    else
+    {
+        aai_result = TryReadViaAccessInterface(
+            request.path, GetAttributeAccessOverride(request.path.mEndpointId, request.path.mClusterId), encoder);
+    }
+    ReturnErrorCodeIf(aai_result.has_value(), *aai_result);
+
+    if (!std::holds_alternative<const EmberAfAttributeMetadata *>(metadata))
+    {
+        // if we only got a cluster, this was for a global attribute. We cannot read ember attributes
+        // at this point, so give up (although GlobalAttributeReader should have returned something here).
+        chipDie();
+    }
+    const EmberAfAttributeMetadata * attributeMetadata = std::get<const EmberAfAttributeMetadata *>(metadata);
+
+    // At this point, we have to use ember directly to read the data.
+    EmberAfAttributeSearchRecord record;
+    record.endpoint                            = request.path.mEndpointId;
+    record.clusterId                           = request.path.mClusterId;
+    record.attributeId                         = request.path.mAttributeId;
+    Protocols::InteractionModel::Status status = emAfReadOrWriteAttribute(
+        &record, &attributeMetadata, gEmberAttributeIOBufferSpan.data(), static_cast<uint16_t>(gEmberAttributeIOBufferSpan.size()),
+        /* write = */ false);
+
+    if (status != Protocols::InteractionModel::Status::Success)
+    {
+        return ChipError(ChipError::SdkPart::kIMGlobalStatus, to_underlying(status), __FILE__, __LINE__);
```

Comment:
This leaks file/line into binary. Suggest avoiding at least __FILE__

Reply:
Updated to use macros.

Reply:
Macros fail with `chip::Protocols::InteractionModel::Status status` is not const`

because our CHIP_SDK_ERROR are constexpr. I will have to see about a workaround.

Reply:
I added a separate macro that depends on error source config flags for this situation.